packages/rs-dpp/src/data_contract/data_contract.rs

```diff
@@ -85,41 +95,246 @@ impl Convertible for DataContract {
+#[platform_deserialize_from(DataContractInner)]
+#[serde(try_from = "DataContractInner")]
 #[serde(rename_all = "camelCase")]
 pub struct DataContract {
+    /// The version of the protocol this contract adheres to.
+    #[serde(default = "default_protocol_version")]
     pub protocol_version: u32,
+
+    /// A unique identifier for the data contract.
     #[serde(rename = "$id")]
     pub id: Identifier,
+
+    /// A reference to the JSON schema that defines the contract.
     #[serde(rename = "$schema")]
     pub schema: String,
+
+    /// The version of this data contract.
     pub version: u32,
+
+    /// The identifier of the contract owner.
     pub owner_id: Identifier,
+
+    /// A mapping of document names to their corresponding document types.
     #[serde(skip)]
     pub document_types: BTreeMap<DocumentName, DocumentType>,
+
+    /// Optional metadata associated with the contract.
     #[serde(skip)]
     pub metadata: Option<Metadata>,
+
+    /// Internal configuration for the contract.
     #[serde(skip)]
-    pub(crate) config: ContractConfig,
+    pub config: ContractConfig,
 
-    #[serde(rename = "documents")]
+    /// A mapping of document names to their corresponding JSON schemas.
     pub documents: BTreeMap<DocumentName, JsonSchema>,
 
-    // TODO we may ensure in compile time that defs are not empty if we define a type for it
+    /// Optional mapping of definition names to their corresponding JSON schemas.
     #[serde(rename = "$defs", default)]
     pub defs: Option<BTreeMap<DefinitionName, JsonSchema>>,
 
+    /// A randomly generated value used for creating unique identifiers within the contract.
     #[serde(skip)]
     pub entropy: Bytes32,
 
+    /// A nested mapping of document names and property paths to their binary values.
     #[serde(skip)]
     pub binary_properties: BTreeMap<DocumentName, BTreeMap<PropertyPath, JsonValue>>,
 }
 
+impl Encode for DataContract {
+    fn encode<E: Encoder>(&self, encoder: &mut E) -> Result<(), EncodeError> {
+        let inner: DataContractInner = self.clone().into();
+        inner.encode(encoder)
+    }
+}
+
+impl Decode for DataContract {
+    fn decode<D: Decoder>(decoder: &mut D) -> Result<Self, DecodeError> {
+        let inner = DataContractInner::decode(decoder)?;
+        inner
+            .try_into()
+            .map_err(|e: ProtocolError| DecodeError::custom(e.to_string()))
+    }
+}
+
+impl<'a> BorrowDecode<'a> for DataContract {
+    fn borrow_decode<D: BorrowDecoder<'a>>(decoder: &mut D) -> Result<Self, DecodeError> {
+        let inner = DataContractInner::decode(decoder)?;
+        inner
+            .try_into()
+            .map_err(|e: ProtocolError| DecodeError::custom(e.to_string()))
+    }
+}
+
+// Standalone default_protocol_version function
+fn default_protocol_version() -> u32 {
+    1
+}
+
+#[derive(Serialize, Deserialize, Encode, Decode)]
+#[serde(rename_all = "camelCase")]
+pub struct DataContractInner {
```

Comment:
I don't think I fully get the purpose of separating the DataCOntract and DataContractInner. Is it only to get rid of the skipped fields?

Reply:
Yes. Bincode doesn't recognize serde(skip)